[Ready] Skillcapes II: Revenge of the broken github: A Skill Menu Special: New Sprites Edition

MrDoomBringer opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
**Edit: I un-lazied myself and made some new sprites!**
![image](https://i.imgur.com/eEyJDiJ.gif)

[old sprites](https://i.imgur.com/qAJUJjQ.gif)
Adds SKILLCAPES! Finally, a status symbol to show off your incredible gaming/cleaning/mining/healing prowess. These beautiful status symbols are given to you once your skill level reaches Legendary. They include sick animated sprites and ego-boosting descriptions, not to mention belligerently high levels of bragging rights.

Also, slightly improves skill message code to be more modular and much more helpful at indicating what level you are.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Adds more incentives to level up your skill.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: MrDoomBringer
add: Reaching legendary levels of any given skill will grant you an otherwise unobtainable reward.
add: Nanotrasen has recently acquired new software to track employee skill development, straight from your PDA! Check it out! In unrelated news, the rate of summary executions handed out to Nanotrasen's lowest performing employees has risen sharply.
admin: Modify player's skills with the new Skill Panel menu, available from the Player Panel!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Original PR closed due to github issues yadda yadda you know the story

new from lego: skill menus
![image](https://user-images.githubusercontent.com/29008542/81638888-59697280-93e8-11ea-9de4-5ffce760bdc6.png)

![image](https://user-images.githubusercontent.com/29008542/81638896-5c646300-93e8-11ea-9df5-21cccd6f91ce.png)

MrDoomBringer commented:
@Rohesie @ExcessiveUseOfCobblestone

MrDoomBringer commented:
Note that this is a clone of #50656, which has already been approved by @Twaticus

tmtmtl30 commented:
that's this PR

nemvar commented on code/datums/skills/_skill.dm:
```diff
@@ -16,10 +40,35 @@
+	if (new_level != SKILL_LEVEL_LEGENDARY)
+		return
+	if (!ispath(skillcapepath))	
+		to_chat(mind.current, "<span class='nicegreen'>My legendary [name] skill is quite impressive, thoguh it seems the Professional [title] Association doesn't have any status symbols to commemorate my abilities with. I should let Centcom know of this travesty, maybe they can do something about it.</span>")
```
```suggestion
		to_chat(mind.current, "<span class='nicegreen'>My legendary [name] skill is quite impressive, though it seems the Professional [title] Association doesn't have any status symbols to commemorate my abilities with. I should let Centcom know of this travesty, maybe they can do something about it.</span>")

```

AnturK commented:
Not a fan of these "randomly drops out of the sky in a pod" deliveries.

Rohesie commented:
Twat approved the art, to be clear.
And what you mean is that this is a clone of #50362

Rohesie commented on code/datums/skills/_skill.dm:
```diff
@@ -1,10 +1,34 @@
+  *
+  *Can't happen in the datum's definition because these lists are not constant expressions
+  */
+/datum/skill/New()
+	. = ..()
+	levelUpMessages = list("<span class='nicegreen'>I'm starting to figure out what [name] really is!</span>",
+	"<span class='nicegreen'>I'm getting a little better at [name]!</span>",
+	"<span class='nicegreen'>I'm getting much better at [name]!</span>",
+	"<span class='nicegreen'>I feel like I've become quite proficient at [name]!</span>",
+	"<span class='nicegreen'>After lots of practice, I've begun to truly understand the intricies \
+	 and surprising depth behind [name]. I now condsider myself a master [title].</span>",
+	"<span class='nicegreen'>Through incredible determination and effort, I've reached the peak of my [name] abiltities. I'm finally able to consider myself a legendary [title]!</span>" )
+	levelDownMessages = list("<span class='nicegreen'>I'm starting to forget what [name] really even is. I need more practice...</span>",
+	"<span class='nicegreen'>I'm getting a little worse at [name]. I'll need to keep practicing to get better at it...</span>",
+	"<span class='nicegreen'>I'm getting a little worse at [name]...</span>",
+	"<span class='nicegreen'>I'm losing my [name] expertise ....</span>",
+	"<span class='nicegreen'>I feel like I'm losing my mastery of [name].</span>",
+	"<span class='nicegreen'>I feel as though my legendary [name] skills have deteriorated. I'll need more intense training to recover my lost skills.</span>" )
```
Put these in the list var definitions.

MrDoomBringer replied:
Do you mean putting these list definitions on line 7 and 8? 
![image](https://user-images.githubusercontent.com/29008542/80250903-adcbce80-8642-11ea-8924-93aa0ed2566a.png)

MrDoomBringer replied:
Because the lists are non-constant expressions and shouldn't go there

Rohesie replied:
```cpp
var/levelUpMessages = list("<span class='nicegreen'>I'm starting to figure out what [name] really is!</span>",
	"<span class='nicegreen'>I'm getting a little better at [name]!</span>",
	"<span class='nicegreen'>I'm getting much better at [name]!</span>",
	"<span class='nicegreen'>I feel like I've become quite proficient at [name]!</span>",
	"<span class='nicegreen'>After lots of practice, I've begun to truly understand the intricies \
	 and surprising depth behind [name]. I now condsider myself a master [title].</span>",
	"<span class='nicegreen'>Through incredible determination and effort, I've reached the peak of my [name] abiltities. I'm finally able to consider myself a legendary [title]!</span>" )
```
What benefit are you aiming for by doing it in `Initialize()`?

MrDoomBringer replied:
uuh its literally impossible to put them in the list definitions because they are non-constant expressions

MrDoomBringer replied:
unless i'm missing something

Rohesie replied:
You can do it, try it.

MrDoomBringer replied:
![image](https://user-images.githubusercontent.com/29008542/80254850-e8853500-8649-11ea-9fd9-3d75fc32c102.png)

Rohesie replied:
Name it `var/list/levelUpMessage`

MrDoomBringer replied:
What? You know that doesn't change it from being a non-constant expression, right?

MrDoomBringer replied:
(typing it as a list doesn't change anything, btw)

MrDoomBringer replied:
(just to be sure, i gave it an explicit list type and it didn't change anything)

Qustinnus replied:
rohesie smoked too much of the weed, put it in init

Rohesie replied:
Right, it would have helped if I actually looked at the contents of the list.
Ignore everything I said.

ExcessiveUseOfCobblestone commented:
> 
> 
> Not a fan of these "randomly drops out of the sky in a pod" deliveries.

maybe a craftable /w skill req?

Angustmeta commented on code/datums/skills/gaming.dm:
```diff
@@ -1,15 +1,13 @@
+	. = ..()
+	levelUpMessages[1] = "<span class='nicegreen'>I'm starting to get a hang of the controls of these games...</span>"
+	levelUpMessages[4] = "<span class='nicegreen'>I'm starting to pick up the meta of these arcade games. If I were to minmax the optimal strat and accentuate my playstyle around well-refined tech...</span>"
+	levelUpMessages[6] = "<span class='nicegreen'>Through incredible determination and effort, I've reached the peak of my [name] abiltities. I wonder how I can become any more powerful... maybe gamerfuel would actually would help me play better..?</span>"
```
```suggestion
	levelUpMessages[6] = "<span class='nicegreen'>Through incredible determination and effort, I've reached the peak of my [name] abilities. I wonder how I can become any more powerful... Maybe gamer fuel would actually help me play better..?</span>"
```

MrDoomBringer replied:
thank you i cannot spell

MrDoomBringer commented:
> Not a fan of these "randomly drops out of the sky in a pod" deliveries.

i think its funny and cool so grr bark

Qustinnus commented:
whyd u ping antur tho

MrDoomBringer commented:
@Qustinnus cus he's the one i quoted u chumbis
![image](https://user-images.githubusercontent.com/29008542/80260741-91399180-8656-11ea-8599-0fccca8f3030.png)

Qustinnus commented:
why u pinging me i didnt say shit dont @ me ill flay u :^)

MrDoomBringer commented:
![image](https://user-images.githubusercontent.com/29008542/80260894-ed9cb100-8656-11ea-8cc8-fa73a8a832d4.png)
nice try dum idiot but github comments take a sec to delete ;^)

Rohesie commented:
Code-wise this is fine. Just needs the concerns of AnturK addressed.

ATH1909 commented:
> ![image](https://user-images.githubusercontent.com/29008542/80260894-ed9cb100-8656-11ea-8cc8-fa73a8a832d4.png)
> nice try dum idiot but github comments take a sec to delete ;^)

qustinnus exposed

imsxz commented:
I don’t necessarily see the issue with drop pod deliveries? Your complaint doesn’t seem context specific so I’ll assume you mean in general. There’s not many ways of quick convenient delivery that will make sense short of an amazon drone delivery.

optimumtact commented:
just make them a craftable it's not rocket science

MrDoomBringer commented:
@optimumtact see my earlier comment addressing those two points uwu

MrDoomBringer commented:
see my earlier comment addressing those two points uwu

optimumtact commented:
You're confusing me with someone who cares. Either work it out with Anturk or do as they say.

MrDoomBringer commented:
>makes a comment on this PR outlining his opinions on how to carry out some changes
>n-not like i care or anything

😏 

anyways @AnturK are you saying this shouldn't be merged until I change the delivery system or are you tolerating it for now?

Rohesie commented:
> > makes a comment on this PR outlining his opinions on how to carry out some changes
> > n-not like i care or anything
> 
> 😏
> 
> anyways @AnturK are you saying this shouldn't be merged until I change the delivery system or are you tolerating it for now?

Not very nice.
Cobby and oranges provided two possible solutions. Take your pick or find a different one.

MrDoomBringer commented:
AnturK hasn't responded on what his preference would be. I'll wait to make a decision until I'm sure that he'll be happy with it moving forward, that's all.

AnturK commented:
I was thinking of some dedicated pickup machine for personal centcom mail/packages. I could see it used for more things like these.

MrDoomBringer commented:
https://cdn.discordapp.com/attachments/500382796334759936/703853839828385812/unknown.png
Fleshing out a skill monitoring / reward PDA program after some spitballing with AnturK

Qustinnus commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -356,7 +357,28 @@ GLOBAL_LIST_EMPTY(PDAs)
+				dat += "<br><br>This software is designed to track and monitor your expertise as a NanoTrasen employee. Your performance in different areas of expertise has been categorized below.<br><HR>"
+				for (var/type in subtypesof(/datum/skill))
+					var/datum/skill/S = GetSkillRef(type)
+					var/skill_level_num = user.mind.get_skill_level(S)+1
```
why plus one

MrDoomBringer replied:
cus skill_level returns 0-6 and level_names is a list with indexes 1-7 chumbis

Qustinnus commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -356,7 +357,28 @@ GLOBAL_LIST_EMPTY(PDAs)
+					var/skill_level_name = uppertext(SSskills.level_names[skill_level_num])
+					var/skill_exp = user.mind.skill_experience[S]
+					var/required_exp = user.mind.get_xp_to_level_up(S)
+					var/exp_percent = skill_exp/required_exp
```
this is a weird way to do it; this'll show progress from level 0 to the next level; instead you should compared the difference between your experience minus the experience needed for the previous level divided between the difference between the previous level and next level,

MrDoomBringer replied:
oops i meant to do it that way im just retarded

Qustinnus commented on code/datums/mind.dm:
```diff
@@ -125,6 +125,20 @@
+		return SKILL_EXP_EXPERT
+	if (exp < SKILL_EXP_MASTER)
+		return SKILL_EXP_MASTER
+	return -1	
```
make this return 0 instead

MrDoomBringer replied:
ok

Qustinnus commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -356,7 +357,28 @@ GLOBAL_LIST_EMPTY(PDAs)
+					dat += "<br>"
+					dat += "<ul><li>"
+					dat += "<br>CURRENT EMPLOYEE LEVEL OF EXPERTISE: [skill_level_name]"
+					if (required_exp != -1)
```
combined with my other comment, this can just be if(required_exp)

MrDoomBringer replied:
ur genius is showing

ThePainkiller commented:
As someone who strives to reach Legendary tier on medicine every shift, this is an even bigger incentive than just faster surgery. That medical cloak is absolutely baller and I **NEED** it.

Qustinnus commented on code/__DEFINES/skills.dm:
```diff
@@ -1,27 +1,24 @@
-#define SKILL_LEVEL_EXPERT 4
-#define SKILL_LEVEL_MASTER 5
-#define SKILL_LEVEL_LEGENDARY 6
+#define S_NONE 1
```
this is undescriptive and will not help someone understand what it means

Qustinnus commented on code/datums/mind.dm:
```diff
@@ -125,95 +124,83 @@
-	if (exp > SKILL_EXP_NOVICE)
-		return SKILL_EXP_NOVICE
-	return 0
+/datum/mind/proc/init_known_skills()
```
this proc entirely ignores the reason why this was done like this and floods every mob with a list they might not need, don't do this.

MrDoomBringer replied:
im gonna flood you with a list you dont need

Rohesie replied:
Any reason not to cache these lists?

Qustinnus commented on code/datums/skills/_skill.dm:
```diff
@@ -6,9 +6,10 @@
 	var/skillcapepath
 	var/list/levelUpMessages = list()
 	var/list/levelDownMessages = list()
+	var/list/level_names = list("None", "Novice", "Apprentice", "Journeyman", "Expert", "Master", "Legendary") //List of skill level names. Note that indexes can be accessed like so: level_names[S_NOVICE]
```
this is supposed to be on the subsystem

Qustinnus commented on code/datums/skills/gaming.dm:
```diff
@@ -5,6 +5,7 @@
 	modifiers = list(SKILL_PROBS_MODIFIER = list(0, 5, 10, 15, 15, 20, 25),
 				SKILL_RANDS_MODIFIER = list(0, 1, 2, 3, 4, 5, 7))
 	skillcapepath = /obj/item/clothing/neck/cloak/skill_reward/gaming
+	level_names = list("None", "Gaming Noob", "Casual Gamer", "Intermediate Gamer", "Advanced Gamer", "Pro Gamer", "Gamer Legend") //List of skill level names. Note that indexes can be accessed like so: level_names[S_NOVICE]
```
when I made the system Ninja and me intentionally wanted a df inspired system that doesn't have meme titles for every skill because it makes it more confusing.

Qustinnus commented:
some of these refactored changes are good but a lot of them seem like design changes.

stylemistake commented on tgui/bin/tgui-build.bat:
```diff
@@ -2,4 +2,4 @@
 call yarn install
 call yarn run build
-timeout /t 9
+timeout /t 93
```
LUL

MrDoomBringer replied:
ayo chill shh

stylemistake commented on code/modules/admin/skill_panel.dm:
```diff
@@ -27,12 +27,12 @@ force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.adm
+			xp_req_to_level = SKILL_EXP_LIST[lvl_num+1] - SKILL_EXP_LIST[lvl_num]
 		var/exp_percent = exp / SKILL_EXP_LIST[S_LEGENDARY]
-		.["skills"] += list(list("playername" = targetmind.current, "path" = type, "name" = S.name, "desc" = S.desc, "lvl" = lvl_name, "exp" = exp, "exp_prog" = exp_progress_to_next_level, "exp_req" = exp_required_for_next_level, "exp_percent" = exp_percent))
+		.["skills"] += list(list("playername" = targetmind.current, "path" = type, "name" = S.name, "desc" = S.desc, "lvl" = lvl_name, "exp" = exp, "exp_prog" = xp_req_to_level - xp_prog_to_level, "exp_req" = xp_req_to_level, "exp_percent" = exp_percent))
```
I always admire these long lines.

stylemistake reviewed (commented):
Code wise it's fine, I can nitpick about code style all day long, but capes are awesome, and admin tools for seeing/managing levels are also great.

P.S. You also need to rebase your branch and rebuild tgui to make it mergeable.

MrDoomBringer commented:
@stylemistake thanks, and yeah i'll rebase when I'm done polishing up the PR

@Qustinnus i'm gonna make some some design changes regarding the way your face looks!!

MrDoomBringer commented:
![image](https://user-images.githubusercontent.com/29008542/80597322-ab2bf900-89f5-11ea-904b-ebd97db2ec9c.png)
skill management ui

Rohesie commented on code/modules/vehicles/vehicle_key.dm:
```diff
@@ -55,7 +55,7 @@
 		user.remove_atom_colour(ADMIN_COLOUR_PRIORITY)
 		user.visible_message("<span class='suicide'>[user] forgot [user.p_they()] isn't actually a janicart! That's a paddlin'!</span>")
-		if(user.mind.get_skill_level(/datum/skill/cleaning) >= SKILL_LEVEL_LEGENDARY) //Janny janny janny janny janny
+		if(user.mind.get_skill_level(/datum/skill/cleaning) >=SKILL_LEVEL_LEGENDARY) //Janny janny janny janny janny
```
```suggestion
		if(user.mind.get_skill_level(/datum/skill/cleaning) >= SKILL_LEVEL_LEGENDARY) //Janny janny janny janny janny
```
Why would you make a change like this?

MrDoomBringer replied:
Happened inadvertently when I was doing a ctrl+f find/replace, looks like i didn't catch 2 of these. Oops!

Rohesie commented on code/modules/modular_computers/file_system/programs/arcade.dm:
```diff
@@ -122,7 +122,7 @@
 				healamt = rand(6,8) + rand(0, gamerSkill)
 				var/maxPointCost = 3
-				if(gamerSkillLevel >= SKILL_LEVEL_JOURNEYMAN)
+				if(gamerSkillLevel >=SKILL_LEVEL_JOURNEYMAN)
```
```suggestion
				if(gamerSkillLevel >= SKILL_LEVEL_JOURNEYMAN)
```
What's with this changes? Whywouldyouchangethecodetohavenospaces?

Rohesie commented on code/modules/clothing/suits/cloaks.dm:
```diff
@@ -89,3 +89,41 @@
+	if (user.mind?.get_skill_level(associated_skill_path) < SKILL_LEVEL_LEGENDARY)
+		to_chat(user, "<span class = 'notice'>You feel completely and utterly unworthy to even touch \the [src].</span>")
+		return
+	. = ..()
```
```suggestion
	return ..()
```

Rohesie commented on code/modules/admin/topic.dm:
```diff
@@ -1545,6 +1545,26 @@
+
+		var/target = locate(href_list["skill"])
+		var/datum/mind/D
+		if (istype(target, /mob))
```
```suggestion
		if(ismob(target))
```

MrDoomBringer replied:
am retarded

Rohesie commented on code/modules/admin/skill_panel.dm:
```diff
@@ -0,0 +1,54 @@
+		.["skills"] += list(list("playername" = targetmind.current, "path" = type, "name" = S.name, "desc" = S.desc, "lvl" = lvl_name, "exp" = exp, "exp_prog" = xp_req_to_level - xp_prog_to_level, "exp_req" = xp_req_to_level, "exp_percent" = exp_percent))
+
+/datum/skill_panel/ui_act(action, params)
+	if(..())
+		return
```
```suggestion
	. = ..()
	if(.)
		return
```

Rohesie commented on code/modules/admin/skill_panel.dm:
```diff
@@ -0,0 +1,54 @@
+			var/skill = text2path(params["skill"])
+			var/max_skill = length(SKILL_EXP_LIST)
+			var/number = input("Please insert the number between 0 and [max_skill] corresponding to the level you'd like to set the player to.") as num|null
+			if (number && number > 0 && number <= max_skill )
```
```suggestion
			if (number > 0 && number <= max_skill )
```

Rohesie commented on code/modules/admin/skill_panel.dm:
```diff
@@ -0,0 +1,54 @@
+	var/datum/mind/targetmind
+	var/client/holder //client of whoever is using this datum
+
+/datum/skill_panel/New(H, datum/mind/mind)//H can either be a client or a mob due to byondcode(tm)
+	targetmind = mind
+	if (istype(H,/client))
+		var/client/C = H
+		holder = C //if its a client, assign it to holder
+	else
+		var/mob/M = H
```
Could you use more informative variable names than `H`, `C` and `M`?

Rohesie commented on code/modules/admin/admin.dm:
```diff
@@ -738,22 +739,35 @@
+	set desc = "Edit mobs's experience and skill levels"
+	set name = "Show Skill Panel"
+	var/datum/mind/D
+	if (istype(target, /mob))
```
```suggestion
	if(ismob(target,))
```

Rohesie commented on code/game/objects/items/gift.dm:
```diff
@@ -105,7 +105,7 @@ GLOBAL_LIST_EMPTY(possible_gifts)
 		for(var/V in gift_types_list)
 			var/obj/item/I = V
-			if((!initial(I.icon_state)) || (!initial(I.item_state)) || (initial(I.item_flags) & ABSTRACT))
+			if((!initial(I.icon_state)) || (!initial(I.item_state)) || (initial(I.item_flags) & ABSTRACT) )
```
```suggestion
			if((!initial(I.icon_state)) || (!initial(I.item_state)) || (initial(I.item_flags) & ABSTRACT))
```
If the only change in the file is a single space, then something is up.

MrDoomBringer replied:
it's cus i'm dumb

Rohesie commented on code/game/objects/items/devices/PDA/PDA.dm:
```diff
@@ -356,7 +357,30 @@ GLOBAL_LIST_EMPTY(PDAs)
+				dat += "<h4>[PDAIMG(mail)] ExperTrack® V4.26.2</h4>"
+				dat += "<i>Thank you for choosing ExperTrack® brand software! ExperTrack® inc. is proud to be a NanoTrasen employee expertise and effectiveness department subsidary!</i>"
+				dat += "<br><br>This software is designed to track and monitor your expertise as a NanoTrasen employee. Your performance in different areas of expertise has been categorized below.<br><HR>"
+				for (var/type in subtypesof(/datum/skill))
```
This should be cached into a global list.

Rohesie commented on code/game/objects/items.dm:
```diff
@@ -824,8 +824,8 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
 
-		if(H.mind.get_skill_level(/datum/skill/mining) >= SKILL_LEVEL_JOURNEYMAN && prob(H.mind.get_skill_modifier(/datum/skill/mining, SKILL_PROBS_MODIFIER))) // we check if the skill level is greater than Journeyman and then we check for the probality for that specific level.
-			mineral_scan_pulse(get_turf(H), SKILL_LEVEL_JOURNEYMAN - 2) //SKILL_LEVEL_JOURNEYMAN = 3 So to get range of 1+ we have to subtract 2 from it,.
+		if(H.mind.get_skill_level(/datum/skill/mining) >=SKILL_LEVEL_JOURNEYMAN && prob(H.mind.get_skill_modifier(/datum/skill/mining, SKILL_PROBS_MODIFIER))) // we check if the skill level is greater than Journeyman and then we check for the probality for that specific level.
+			mineral_scan_pulse(get_turf(H),SKILL_LEVEL_JOURNEYMAN - 2) //S_JOURNEYMAN = 3 So to get range of 1+ we have to subtract 2 from it,.
```
```suggestion
		if(H.mind.get_skill_level(/datum/skill/mining) >= SKILL_LEVEL_JOURNEYMAN && prob(H.mind.get_skill_modifier(/datum/skill/mining, SKILL_PROBS_MODIFIER))) // we check if the skill level is greater than Journeyman and then we check for the probality for that specific level.
			mineral_scan_pulse(get_turf(H), SKILL_LEVEL_JOURNEYMAN - 2) //S_JOURNEYMAN = 3 So to get range of 1+ we have to subtract 2 from it,.
```
Please,stopremovingspaces.
Also, why did you change the comment here?

Rohesie commented on code/datums/skills/_skill.dm:
```diff
@@ -1,10 +1,36 @@
+	var/title = "Skiller"
 	var/desc = "the art of doing things"
 	var/modifiers = list(SKILL_SPEED_MODIFIER = list(1, 1, 1, 1, 1, 1, 1)) //Dictionary of modifier type - list of modifiers (indexed by level). 7 entries in each list for all 7 skill levels.
+	var/skillcapepath
```
```suggestion
	var/skill_cape_path
```
Also document the new variables.

Rohesie commented on code/datums/mind.dm:
```diff
@@ -67,118 +67,140 @@
-		if(known_skills[i]) //Do we actually have a level in this?
+	init_known_skills() 
+ 
+/datum/mind/Destroy() 
+	SSticker.minds -= src 
+	if(islist(antag_datums)) 
+		QDEL_LIST(antag_datums) 
+	return ..() 
+ 
+/datum/mind/proc/get_language_holder() 
+	if(!language_holder) 
+		language_holder = new (src) 
+	return language_holder 
```
You seem to be adding a space to every line for some reason, increasing the diff.
Can you fix that?
```suggestion
/datum/mind/Destroy()
	SSticker.minds -= src
	if(islist(antag_datums))
		QDEL_LIST(antag_datums)
	return ..()

/datum/mind/proc/get_language_holder()
	if(!language_holder)
		language_holder = new (src)
	return language_holder
```
Same all around the file.

Rohesie replied:
This is still a thing.

ThePainkiller replied:
Doom, why...would you put spaces? I mean, it's a small detail but these look more intentional than just a typo...I'm confus

Rohesie replied:
This is not an intentional change. It happens with EoL changes.

MrDoomBringer replied:
Fixed it, sorry - didn't notice your comments until just now

Rohesie commented on code/controllers/subsystem/skills.dm:
```diff
@@ -20,4 +19,4 @@ SUBSYSTEM_DEF(skills)
 	for(var/type in subtypesof(/datum/skill))
 		var/datum/skill/ref = new type
-		all_skills[type] = ref
+		all_skills[type] = ref
```
```suggestion
		all_skills[type] = ref

```
Always end on an empty line, else the linter will scream.

Rohesie commented on tgui/packages/tgui/interfaces/SkillPanel.js:
```diff
@@ -21,7 +21,7 @@ export const SkillPanel = (props, context) => {
                 )}
                 {skill.exp_req === 0 && (
-                  "XP to next level: [MAXXED]"
+                  "XP to nexts level: [MAXXED]"
```
```suggestion
                  "XP to next's level: [MAXXED]"
```

MrDoomBringer replied:
PR isn't quite ready for review yet (though it is close), as i'm just committing the changes I make as i go. This particular one was just to test if my tgui yarn dev server was running, as I'm in the process of reworking SkillPanel.js

I'll make a comment and remove the WIP tag when the code's ready to have a look taken at it

MrDoomBringer commented:
@Rohesie Alright, sorry for the wait - just finished with finals week. PR should be ready for review.

ThePainkiller commented:
I'm SO HYPED FOR THIS, finally being able to see skills progress AND gaining a sick cape out of it? This is the dream

stylemistake commented:
Address travis/linter errors, rebuild tgui by running tgui-build.bat once.

MrDoomBringer committed:
10K hours

60,000 minutes
hahah


sure if you want a trailing newline i can do that


default cloak gets default skill


lazy list and no equip good


necessary ig


Update code/datums/skills/_skill.dm

Co-Authored-By: nemvar <47324920+nemvar@users.noreply.github.com>
Update code/datums/skills/gaming.dm

Co-Authored-By: Angust <46400996+Angustmeta@users.noreply.github.com>
WIP PDA skill monitor program


hacky fixes till i refactor skills


refactors skills a bit, adds admin skill edit menu


If you have a failing, it's that you're always demanding perfection

...IF you have a failing
i think that's it for the admin skill manger


appeases lord flord


bruh


level names


FUCK!!


unga


Update code/modules/vehicles/vehicle_key.dm

Co-authored-by: Rohesie <rohesie@gmail.com>
Update code/modules/admin/topic.dm

Co-authored-by: Rohesie <rohesie@gmail.com>
Update code/modules/admin/skill_panel.dm

Co-authored-by: Rohesie <rohesie@gmail.com>
Update code/modules/admin/skill_panel.dm

Co-authored-by: Rohesie <rohesie@gmail.com>
Apply suggestions from code review

frick

Co-authored-by: Rohesie <rohesie@gmail.com>
EOL and dumb spaces


more rohsie bait


tgui: Query Windows Registry for BYOND cache directory

hell yeah brother


update tgui and dmi


CLRF and bat file


typo ig

MrDoomBringer committed:
tgui and new janicloak that i accdiently changed on another branch

MrDoomBringer commented:
![image](https://i.imgur.com/jDnfXX5.gif)
upgraded miner cloak too cus why not

MrDoomBringer committed:
jani

MrDoomBringer committed:
gamer cloak

MrDoomBringer commented:
Upgraded all the sprites, the gamer sprite now has a sick power game soda background

MrDoomBringer commented:
no YOU'RE cool stuff 😠 ❤️

Rohesie commented:
Do please fix the EoL issues.

MrDoomBringer committed:
trim trailing whitespace

MrDoomBringer committed:
tgui

MrDoomBringer committed:
bruh

MrDoomBringer commented:
EOL issues should be fixed, as well as some variable autodoccing

Rohesie commented on code/modules/admin/admin.dm:
```diff
@@ -739,22 +740,35 @@
-	if(!M.mind)
+	var/datum/mind/D = M.mind
+	if(!D)
 		to_chat(usr, "This mob has no mind!", confidential = TRUE)
 		return
-
-	M.mind.traitor_panel()
+	if(!istype(M) && !istype(D))
+		to_chat(usr, "This can only be used on instances of type /mob and /mind", confidential = TRUE)
+		return
+	D.traitor_panel()
```
Try using a more informative var name than `D`

Rohesie commented on code/modules/admin/admin.dm:
```diff
@@ -739,22 +740,35 @@
+	set name = "Show Skill Panel"
+	var/datum/mind/D
+	if(ismob(target))
+		var/mob/M = target
+		D = M.mind
+	else if (istype(target, /datum/mind))
+		D = target
```
Use something more informative than `M` and `D`

Rohesie commented on code/modules/admin/topic.dm:
```diff
@@ -1545,6 +1545,26 @@
+			return
+
+		var/target = locate(href_list["skill"])
+		var/datum/mind/D
+		if(ismob(target))
+			var/mob/M = target
+			D = M.mind
+		else if (istype(target, /datum/mind))
+			D = target
+		else
+			to_chat(usr, "This can only be used on instances of type /mob and /mind", confidential = TRUE)
+			return
+		show_skill_panel(D)
```
Better var names than `M` and `D` please.

Rohesie commented:
Is AnturK's concern addressed?

MrDoomBringer commented:
@Rohesie AnturK's concern has been addressed, yeah

MrDoomBringer committed:
variable renaming

ThePainkiller reviewed (commented):
![image](https://user-images.githubusercontent.com/1294508/82973864-e4349a80-9fae-11ea-9c73-0bedf01f5fbc.png)
Yeah by the way, clicking on the Skills button gives this, wich...well, it's awkward, we aren't supposed to see the whole path, are we?

MrDoomBringer commented on code/datums/skills/_skill.dm:
```diff
@@ -1,10 +1,42 @@
+	"<span class='nicegreen'>I'm getting much better at [name]!</span>",
+	"<span class='nicegreen'>I feel like I've become quite proficient at [name]!</span>",
+	"<span class='nicegreen'>After lots of practice, I've begun to truly understand the intricies \
+	 and surprising depth behind [name]. I now condsider myself a master [title].</span>",
```
There's a PR open to fix the path thing
The typo you're referring to is "intricies" right? oops

ThePainkiller replied:
Wich should be "intricacies", yes, and also "condsider" wich should only be "consider"!
